### PR TITLE
Core/Stats: Implemented multistrike.

### PR DESCRIPTION
**Pull request by fatalaim**
**Changes proposed**: Implemented multistrike with spell auras SPELL_AURA_MOD_MULTISTRIKE_DMG_PCT and SPELL_AURA_MULTISTRIKE. Needs lots of feedback please.

**Target branch(es)**: 6x

**Tests performed**: (Does it build, tested in-game, etc)
Needs tons of testing.

**Known issues and TODO list**:
- [ ] Implement a single multistrike roll for when in pvp combat.

### CODE REVIEW

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -8199,6 +8331,25 @@ void Unit::EnergizeBySpell(Unit* victim, uint32 spellId, int32 damage, Powers po
...
 }
 
+uint32 Unit::CalculateMultistrikeAmount(Unit* victim, int32 baseAmount)
+{
+    // Player totems benefit from player auras
+    if (IsTotem() && GetOwnerGUID().IsPlayer() && GetOwner())
+        return GetOwner()->CalculateMultistrikeAmount(victim, baseAmount);
+
+    // Hack for Storm, Fire and Earth elementals
+    if ((GetEntry() == 77936 || GetEntry() == 15352 || GetEntry() == 15438) && GetOwnerGUID().IsPlayer() && GetOwner())
```
Comment (DDuarte) [inline src/server/game/Entities/Unit/Unit.cpp]: Could we have some further explanations why this is needed?

Comment (fatalaim) [inline src/server/game/Entities/Unit/Unit.cpp]: Elementals can multistrike but I'm not sure if it's handled by pet spells or similar to totems. Once we figure that out, we can change that line appropriately.

Comment (fatalaim) [inline src/server/game/Entities/Unit/Unit.cpp]: Plus if you look in Unit::GetUnitSpellCriticalChance, a similar hack check is used for elementals.

**Comment by Ovahlord** [comment]
The hook for the multistrike is placed at the wrong location just like at many procs before:
the multistrike rolls already when the spell is launching towards the target, not just after hit.

**Comment by Keader** [comment]
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Handlers/ToyHandler.cpp:59:36: fatal error: 
      comparison of integers of different signs: 'int32' (aka 'int') and
      'const uint32' (aka 'const unsigned int') [-Wsign-compare]
        return packet.Cast.SpellID == effect->SpellID;

**Comment by fatalaim** [comment]
@Ovahlord Do you have a source for that? If so, where should the initial roll be?

**Comment by P-Kito** [comment]
@fatalaim I am playing retail currently and when I cast a Chaos Bolt, another multistrike of it will fly over to the target, even when the first one didnt hit yet.

Can provide video.... but you can trust me x)

**Comment by fatalaim** [comment]
Good catch. Tested it on my retail shaman while standing at max range. I'll rewrite this when I get a chance.

**Comment by Aokromes** [comment]
@fatalaim 

/home/travis/build/TrinityCore/TrinityCore/src/server/game/Spells/Spell.cpp:6964:52: fatal error: 
      unused parameter 'e_time' [-Wunused-parameter]
bool SpellDelayedMultistrikeDamage::Execute(uint64 e_time, uint32 p_time)
                                                   ^
[ 42%] Building CXX object src/server/game/CMakeFiles/game.dir/Spells/SpellHistory.cpp.o
[ 42%] Building CXX object src/server/game/CMakeFiles/game.dir/Spells/SpellEffects.cpp.o
[ 42%] Building CXX object src/server/game/CMakeFiles/game.dir/Spells/Auras/SpellAuraEffects.cpp.o
1 error generated.
make[2]: *** [src/server/game/CMakeFiles/game.dir/Spells/Spell.cpp.o] Error 1
[ 42%] Building CXX object src/server/game/CMakeFiles/game.dir/Spells/Auras/SpellAuras.cpp.o
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Spells/Auras/SpellAuraEffects.cpp:6555:36: fatal error: 
      comparison of unsigned expression < 0 is always false
      [-Wtautological-compare]
        if (multiInfo1->overAmount < 0)
            ~~~~~~~~~~~~~~~~~~~~~~ ^ ~

**Comment by Aokromes** [comment]
Anyone can takeover this PR?

**Comment by tkrokli** [comment]
http://wow.gamepedia.com/Multistrike
- WoD Patch 6.0.2 (2014-10-14): Added
- Legion Patch 7.0.3 (Beta/PTR): Removed.
  - All affected items and spells (such as procs) have been converted to give a bonus to Versatility instead.

**Pull request closed without merging**